include/core/federated/network/socket_common.h

```diff
@@ -0,0 +1,230 @@
+ * to open a socket server on this port.
+ */
+#define DEFAULT_PORT 15045u
+
+/**
+ * Byte identifying that the federate or the RTI has failed.
+ */
+#define MSG_TYPE_FAILED 25
+
+/**
+ * Mutex protecting socket close operations.
+ */
+extern lf_mutex_t socket_mutex;
+
+/**
+ * @brief Create an IPv4 TCP socket with Nagle's algorithm disabled
+ * (TCP_NODELAY) and Delayed ACKs disabled (TCP_QUICKACK). Exits application
+ * on any error.
+ *
+ * @return The socket ID (a file descriptor).
+ */
+int create_real_time_tcp_socket_errexit();
+
+/**
+ * Create a TCP or UDP server and enable listening for socket connections.
+ * If the specified port if it is non-zero, it will attempt to acquire that port.
+ * If it fails, it will repeatedly attempt up to PORT_BIND_RETRY_LIMIT times with
+ * a delay of PORT_BIND_RETRY_INTERVAL in between. If the specified port is
+ * zero, then it will attempt to acquire DEFAULT_PORT first. If this fails, then it
+ * will repeatedly attempt up to PORT_BIND_RETRY_LIMIT times, incrementing the port
+ * number between attempts, with no delay between attempts.  Once it has incremented
+ * the port number MAX_NUM_PORT_ADDRESSES times, it will cycle around and begin again
+ * with DEFAULT_PORT.
+ *
+ * @param port The port number to use or 0 to start trying at DEFAULT_PORT.
+ * @param socket_type The type of the socket for the server (TCP or UDP).
+ * @param final_socket The socket descriptor on which to accept connections.
+ * @param final_port The final port of the TCP or UDP socket.
+ */
+void create_TCP_server(uint16_t port, int* final_socket, uint16_t* final_port);
+void create_UDP_server(uint16_t port, int* final_socket, uint16_t* final_port);
+
+/**
+ * These two functions waits for an incoming connection request on the specified server socket.
+ * It blocks until a connection is successfully accepted. If an error occurs that is not
+ * temporary (e.g., `EAGAIN` or `EWOULDBLOCK`), it reports the error and exits. Temporary
+ * errors cause the function to retry accepting the connection. The accept_federate_socket() function additionally
+ * checks the RTI's server socket if it is still alive.
+ *
+ * @param socket The server socket file descriptor that is listening for incoming connections.
+ * @param rti_socket The rti socket for the federate to check if it is still open.
+ * @return int The file descriptor for the newly accepted socket on success, or -1 on failure
+ *             (with an appropriate error message printed).
+ */
+int accept_rti_socket(int socket);
+int accept_federate_socket(int socket, int rti_socket);
+
+/**
+ *
+ * This function attempts to establish a TCP connection to the specified hostname
+ * and port. It uses `getaddrinfo` to resolve the hostname and retries the connection
+ * periodically if it fails. If the specified port is 0, it iterates through a range
+ * of default ports starting from `DEFAULT_PORT`. The function will stop retrying
+ * if the `CONNECT_TIMEOUT` is reached.
```

Comment:
It bothers me that this function uses port number 0 to mean what port number 1 means for the create_socket function. This is potentially confusing.  I think maybe a better design would to have a boolean argument `increment_port_on_retry` for both functions. Then, to get the effect of port number 1, you specify port number `DEFAULT_PORT` and `increment_port_on_retry == true`.

Reply:
I agree with that it could be confusing.
I reverted the changes, and set the RTI initial port back to zero. I added the `increment_port_on_retry.`

It really looks much better now!

Reply:
I also cleaned up all comments about changing the port number in the code, and the PR description.